Allow start/pause of streaming image input. Only access the webcam while it's needed

freddyaboulton opened the pull request:
## Description

Closes: #3642
Closes: #7230

If the webcam is an allowed image source, gradio will access the webcam as soon as the component is mounted on the page. There is also no way to stop gradio from accessing the webcam. This is #3642

When the webcam is an allowed source, users have to click into the image component to grant webcam access. This is what it looks like in that case:

<img width="616" alt="image" src="https://github.com/gradio-app/gradio/assets/41651716/500f17cb-6d51-4f26-b162-d5e610cefa31">

After the webcam stops recording or taking a picture, gradio will stop using the webcam and revert back to the ^previous state.

Additionally, when `gr.Image(streaming=true)`, the webcam component will behave similarly to video recording. There will be a record and stop button at the bottom that users can use to stop the stream of images. 

### Webcam Demo no streaming
![webcam_demo](https://github.com/gradio-app/gradio/assets/41651716/ca89ed99-c75a-4655-a8aa-d37519a29112)

### Webcam Demo with streaming
![stream_frames](https://github.com/gradio-app/gradio/assets/41651716/cef32fd2-86f4-4a10-b686-57abadfcbf8b)

Test streaming with this demo
```python
import gradio as gr
import numpy as np

def flip(im):
    return np.flipud(im)

with gr.Blocks() as demo:
    with gr.Row():
        input_img = gr.Image(sources=["webcam"])
        output_img = gr.Image()
        input_img.stream(flip, input_img, output_img)
demo.launch(share=True)

if __name__ == "__main__":
    demo.launch()
    
```

EDIT:

Also fixes Interface streaming blocking the output, fixes: #7021 

## 🎯 PRs Should Target Issues

Before your create a PR, please check to see if there is [an existing issue](https://github.com/gradio-app/gradio/issues) for this change. If not, please create an issue before you create this PR, unless the fix is very small. 

Not adhering to this guideline will result in the PR being closed. 

## Tests

1. PRs will only be merged if tests pass on CI. To run the tests locally, please set up [your Gradio environment locally](https://github.com/gradio-app/gradio/blob/main/CONTRIBUTING.md) and run the tests: `bash scripts/run_all_tests.sh`

2. You may need to run the linters: `bash scripts/format_backend.sh` and `bash scripts/format_frontend.sh`

freddyaboulton commented on js/image/shared/Webcam.svelte:
```diff
@@ -180,19 +203,34 @@
 		options_open = false;
 	}
+
+	$: console.log("webcam_accessed", webcam_accessed);
```
will delete 🤦

freddyaboulton commented on js/image/shared/Webcam.svelte:
```diff
@@ -180,19 +203,34 @@
+				<WebcamIcon />
+			</span>
+			<button on:click={async () => await access_webcam()}
+				>{"Click to access webcam"}</button
```
This should probably be camel case for consistency

aliabid94 commented:
I could've sworn I made the change to set the progress tracker visibility to False when streaming from a webcam

abidlabs commented:
> I could've sworn I made the change to set the progress tracker visibility to False when streaming from a webcam

I'm pretty sure its a regression. We should do this for any streaming event, e.g. https://github.com/gradio-app/gradio/issues/7230 as well

abidlabs commented:
This looks great to me @freddyaboulton! My only feedback is that the clickable area to give access to the webcam is very small. I kept clicking the icon and was confused why it wasn't working. I'd suggest making this whole area clickable:

<img width="462" alt="image" src="https://github.com/gradio-app/gradio/assets/1778297/f9a5fee8-b284-45cc-a170-e4db207864a4">

aliabid94 commented:
Is the demo supposed to be using an audio component? Btw when I was running locally I did not see the progress bar but also I got this error after a while when streaming a webcam:
`ValueError: Cannot process this value as an Image, it is of type: <class 'tuple'>`

aliabid94 commented:
> Btw when I was running locally I did not see the progress bar

Actually there was a visible progress bar only when using streaming from Interface, fixed in this PR in this commit: https://github.com/gradio-app/gradio/pull/7228/commits/db019e4c2a61f8fe16f68dc915982c060182267d

freddyaboulton commented:
> Actually there was a visible progress bar only when using streaming from Interface, fixed in this PR in this commit: https://github.com/gradio-app/gradio/commit/db019e4c2a61f8fe16f68dc915982c060182267d

Nice - confirmed it works! 🔥 

> Is the demo supposed to be using an audio component?

Yea that's my bad - copy/pasted the wrong demo. I updated the PR description to include the correct one but also including it here

```python
import gradio as gr
import numpy as np

def flip(im):
    return np.flipud(im)

with gr.Blocks() as demo:
    with gr.Row():
        input_img = gr.Image(sources=["webcam"])
        output_img = gr.Image()
        input_img.stream(flip, input_img, output_img)
demo.launch(share=True)

if __name__ == "__main__":
    demo.launch()
```

> My only feedback is that the clickable area to give access to the webcam is very small. 

Agreed - just pushed up a fix! Will wait for CI to pass before opening it up for broader review.

aliabid94 commented on js/image/shared/Webcam.svelte:
```diff
@@ -131,7 +142,20 @@
+		if (mode === "image" && streaming) {
+			recording = !recording;
+		}
+		const func = mode === "image" ? take_picture : take_recording;
+		func();
```
let's avoid abbreviations like `func`. Personally think this is more readable tbh
```suggestion
		if (mode === "image") {
		  take_picture();
		} else {
		  take_recording();
		}
```

aliabid94 commented on js/image/shared/Webcam.svelte:
```diff
@@ -253,6 +282,10 @@
 		height: var(--size-full);
 	}
 
+	.invisible {
```
nit, but we use the class name `hide` throughout the codebase for `display: none;`, for consistency would be nice

aliabid94 reviewed (approved):
LGTM great work

dawoodkhan82 commented:
Looks great @freddyaboulton. Small nit, I notice after taking a picture the input component quickly flashes back to the "click to access webcam" state before showing the taken picture.

abidlabs commented:
Maybe also fixes: https://github.com/gradio-app/gradio/issues/7230?

freddyaboulton commented:
Thanks for the reviews everyone! Will merge the comments and flicker have been addressed.
